### PR TITLE
New REST endpoint: read single bookmark

### PR DESCRIPTION
**Pull request by marcelklehr**
So, in order to add a tag without deleting the already existing ones, you need to know what tags are already set. Unfortunately there's no way to get that info on a single bookmark without fetching them all at once (uff.), ... until now :)

This pr adds a REST endpoint that proxies libBookmarks#findUniqueBookmark and even takes care of public/private bookmarks.

### CODE REVIEW

**Comment by marcelklehr** [comment]
For the record. The tests fail on an unrelated technicality, of which I'm not sure if it's happening on master, too. (here we go again...)

**Comment by blizzz** [comment]
Sorry for the delay. 

The tests fail all for postgre, other tests against master run fine

File: controller/rest/bookmarkcontroller.php [hunk 1]
```diff
@@ -54,6 +54,36 @@ public function __construct($appName, IRequest $request, $userId, IDBConnection
...
+			$publicOnly = false;
+		}else {
+			$publicOnly = true;
+			if ($this->userManager->userExists($user) == false) {
+				$error = "User could not be identified";
+				return new JSONResponse(array('status' => 'error', 'data'=> $error), Http::STATUS_BAD_REQUEST);
+			}
+		}
+		$bm = $this->bookmarks->findUniqueBookmark($id, $user);
+    if ($publicOnly === TRUE && $bm['public'] !== '1') {
```
Comment (blizzz) [inline controller/rest/bookmarkcontroller.php]: the indentation here and in the function declaration is off

**Comment by blizzz** [comment]
There's indeed not much potential what could go wrong… perhaps `$bm['public']` ends up as int?

**Comment by marcelklehr** [comment]
re my comment before: I made that remark in reference to an earlier state of this pr :)

Seems now as if it's a type incoherence, yep.

**Comment by marcelklehr** [comment]
Here we go :)

**Comment by marcelklehr** [comment]
having this in the next release would allow floccus to sync folder structures by applying a tag with the folder path as an intermediate solution.

File: controller/rest/bookmarkcontroller.php [hunk 2]
```diff
@@ -65,21 +65,21 @@ public function legacyGetBookmarks($type = "bookmark", $tag = '', $page = 0, $so
 	 */
 	public function getSingleBookmark(
 		$id,
-    $user = null
+		$user = null
 	) {
 		if ($user === null) {
 			$user = $this->userId;
-			$publicOnly = false;
+			$public_only = false;
```
Comment (blizzz) [inline controller/rest/bookmarkcontroller.php]: no, please don't, camelCase is the coding style https://docs.nextcloud.com/server/12/developer_manual/general/codingguidelines.html#objects-functions-arrays-variables

**Comment by blizzz** [comment]
Let's return 404 for a non-existing bookmark:

```diff
diff --git a/controller/rest/bookmarkcontroller.php b/controller/rest/bookmarkcontroller.php
index bd65bd74..bf0e95b6 100644
--- a/controller/rest/bookmarkcontroller.php
+++ b/controller/rest/bookmarkcontroller.php
@@ -78,7 +78,10 @@ class BookmarkController extends ApiController {
            }
        }
        $bm = $this->bookmarks->findUniqueBookmark($id, $user);
-       if ($public_only === TRUE && $bm['public'] != '1') {
+       if(!isset($bm['id'])) {
+           return new JSONResponse(['status' => 'error'], Http::STATUS_NOT_FOUND);
+       }
+       if ($public_only === true && isset($bm['public']) && $bm['public'] != '1') {
            $error = "Insufficient permissions";
            return new JSONResponse(array('status' => 'error', 'data' => $error), Http::STATUS_BAD_REQUEST);
     }
diff --git a/tests/bookmarkcontroller_test.php b/tests/bookmarkcontroller_test.php
index 3eabbda7..a88000a1 100644
--- a/tests/bookmarkcontroller_test.php
+++ b/tests/bookmarkcontroller_test.php
@@ -65,6 +65,15 @@ class Test_BookmarkController extends TestCase {
        $this->assertEquals("http://www.9gag.com", $data['item']['url']);
    }
 
+   function testPrivateReadNotFound() {
+       $this->cleanDB();
+       $this->setupBookmarks();
+       $output = $this->controller->getSingleBookmark(987);
+       $data = $output->getData();
+       $this->assertSame('error', $data['status']);
+       $this->assertSame(404, $output->getStatus());
+   }
+
   function testPublicReadSuccess() {
        $this->cleanDB();
        $this->setupBookmarks();

```

and get it in.

File: controller/rest/bookmarkcontroller.php (hunk 2 above)
Comment (marcelklehr) [reply controller/rest/bookmarkcontroller.php]: that's what I thought, until I misinterpreted your comment: "keep true lower case"... :laughing:

**Review by blizzz — APPROVED** [review]

**Comment by blizzz** [comment]
thx @marcelklehr :)

**Comment by marcelklehr** [comment]
:tada:

**Pull request merged**